File: tests/test_synthesis.py
import os
from nose.tools import assert_almost_equals
import simpl
import simpl.peak_detection as peak_detection
import simpl.partial_tracking as partial_tracking
import simpl.synthesis as synthesis

float_precision = 2
frame_size = 512
hop_size = 512
max_peaks = 10
max_partials = 10
num_frames = 30
num_samples = num_frames * hop_size
audio_path = os.path.join(
    os.path.dirname(__file__), 'audio/flute.wav'
)
libsms_harmonic_synthesis_ifft_path = os.path.join(
    os.path.dirname(__file__), 'libsms_harmonic_synthesis_ifft.wav'
)
libsms_harmonic_synthesis_sin_path = os.path.join(
    os.path.dirname(__file__), 'libsms_harmonic_synthesis_sin.wav'
)

PeakDetection = peak_detection.PeakDetection
SMSPeakDetection = peak_detection.SMSPeakDetection
PartialTracking = partial_tracking.PartialTracking
SMSPartialTracking = partial_tracking.SMSPartialTracking
Synthesis = synthesis.Synthesis
SMSSynthesis = synthesis.SMSSynthesis


class TestSynthesis(object):
    @classmethod
    def setup_class(cls):
        cls.audio = simpl.read_wav(audio_path)[0]
        cls.audio = cls.audio[0:num_samples]

    def test_basic(self):
        pd = PeakDetection()
        pd.hop_size = hop_size
        frames = pd.find_peaks(self.audio)

        pt = PartialTracking()
        pt.max_partials = max_partials
        frames = pt.find_partials(frames)

        s = Synthesis()
        s.hop_size = hop_size
        synth_audio = s.synth(frames)

        assert len(synth_audio) == len(self.audio),\
            (len(synth_audio), len(self.audio))


class TestSMSSynthesis(object):
    @classmethod
    def setup_class(cls):
        cls.audio = simpl.read_wav(audio_path)[0]
        cls.audio = cls.audio[0:num_samples]

    def test_basic(self):
        pd = SMSPeakDetection()
        pd.hop_size = hop_size
        frames = pd.find_peaks(self.audio)

        pt = SMSPartialTracking()
        pt.max_partials = max_partials
        frames = pt.find_partials(frames)

        s = SMSSynthesis()
        s.hop_size = hop_size
        synth_audio = s.synth(frames)

        assert len(synth_audio) == len(self.audio),\
            (len(synth_audio), len(self.audio))

    def test_harmonic_synthesis_ifft(self):
        pd = SMSPeakDetection()
        pd.hop_size = hop_size
        frames = pd.find_peaks(self.audio)

        pt = SMSPartialTracking()
        pt.max_partials = max_partials
        frames = pt.find_partials(frames)

        synth = SMSSynthesis()
        synth.hop_size = hop_size
        synth.max_partials = max_partials
        synth.det_synthesis_type = SMSSynthesis.SMS_DET_IFFT
        synth_audio = synth.synth(frames)

        assert len(synth_audio) == len(self.audio)

        sms_audio, sampling_rate = simpl.read_wav(
            libsms_harmonic_synthesis_ifft_path
        )

        assert len(synth_audio) == len(sms_audio)

        for i in range(len(synth_audio)):
            assert_almost_equals(synth_audio[i], sms_audio[i], float_precision)

    def test_harmonic_synthesis_sin(self):
        pd = SMSPeakDetection()
        pd.hop_size = hop_size
        frames = pd.find_peaks(self.audio)

        pt = SMSPartialTracking()
        pt.max_partials = max_partials
        frames = pt.find_partials(frames)

        synth = SMSSynthesis()
        synth.hop_size = hop_size
        synth.max_partials = max_partials
        synth.det_synthesis_type = SMSSynthesis.SMS_DET_SIN
        synth_audio = synth.synth(frames)

        assert len(synth_audio) == len(self.audio)

        sms_audio, sampling_rate = simpl.read_wav(
            libsms_harmonic_synthesis_sin_path
        )

        assert len(synth_audio) == len(sms_audio)

        for i in range(len(synth_audio)):
            assert_almost_equals(synth_audio[i], sms_audio[i], float_precision)

File: tests/test_synthesis.cpp
#include "test_synthesis.h"

using namespace simpl;

// ---------------------------------------------------------------------------
//	test_basic
// ---------------------------------------------------------------------------
static void test_basic(PeakDetection *pd, PartialTracking* pt,
                       Synthesis* synth, SndfileHandle *sf) {
    int num_samples = 4096;

    std::vector<sample> audio(sf->frames(), 0.0);
    sf->read(&audio[0], (int)sf->frames());

    pd->clear();
    pt->reset();
    synth->reset();

    Frames frames = pd->find_peaks(num_samples,
                                   &(audio[(int)sf->frames() / 2]));
    frames = pt->find_partials(frames);
    frames = synth->synth(frames);

    for(int i = 0; i < frames.size(); i++) {
        CPPUNIT_ASSERT(frames[i]->num_peaks() > 0);
        CPPUNIT_ASSERT(frames[i]->num_partials() > 0);

        double energy = 0.f;
        for(int j = 0; j < synth->hop_size(); j++) {
            energy += frames[i]->synth()[j] * frames[i]->synth()[j];
        }
        CPPUNIT_ASSERT(energy > 0.f);
    }
}

// ---------------------------------------------------------------------------
//	test_changing_frame_size
// ---------------------------------------------------------------------------
static void test_changing_frame_size(PeakDetection *pd, PartialTracking* pt,
                                     Synthesis* synth, SndfileHandle *sf) {
    int num_samples = 4096;
    int frame_size = 256;
    int hop_size = 128;

    std::vector<sample> audio(sf->frames(), 0.0);
    sf->read(&audio[0], (int)sf->frames());

    pd->clear();
    pt->reset();
    synth->reset();

    pd->frame_size(frame_size);
    pd->hop_size(hop_size);
    synth->frame_size(frame_size);
    synth->hop_size(hop_size);

    Frames frames = pd->find_peaks(num_samples,
                                   &(audio[(int)sf->frames() / 2]));
    frames = pt->find_partials(frames);
    frames = synth->synth(frames);

    for(int i = 0; i < frames.size(); i++) {
        CPPUNIT_ASSERT(frames[i]->num_peaks() > 0);
        CPPUNIT_ASSERT(frames[i]->num_partials() > 0);

        double energy = 0.f;
        for(int j = 0; j < synth->hop_size(); j++) {
            energy += frames[i]->synth()[j] * frames[i]->synth()[j];
        }
        CPPUNIT_ASSERT(energy > 0.f);
    }

    frame_size = 128;
    hop_size = 64;

    pd->clear();
    pt->reset();
    synth->reset();

    pd->frame_size(frame_size);
    pd->hop_size(hop_size);
    synth->frame_size(frame_size);
    synth->hop_size(hop_size);

    frames = pd->find_peaks(num_samples, &(audio[(int)sf->frames() / 2]));
    frames = pt->find_partials(frames);
    frames = synth->synth(frames);

    for(int i = 0; i < frames.size(); i++) {
        CPPUNIT_ASSERT(frames[i]->num_peaks() > 0);
        CPPUNIT_ASSERT(frames[i]->num_partials() > 0);

        double energy = 0.f;
        for(int j = 0; j < synth->hop_size(); j++) {
            energy += frames[i]->synth()[j] * frames[i]->synth()[j];
        }
        CPPUNIT_ASSERT(energy > 0.f);
    }
}

// ---------------------------------------------------------------------------
//	TestMQSynthesis
// ---------------------------------------------------------------------------
void TestMQSynthesis::setUp() {
    _sf = SndfileHandle(TEST_AUDIO_FILE);

    if(_sf.error() > 0) {
        throw Exception(std::string("Could not open audio file: ") +
                        std::string(TEST_AUDIO_FILE));
    }
}

void TestMQSynthesis::test_basic() {
    ::test_basic(&_pd, &_pt, &_synth, &_sf);
}

void TestMQSynthesis::test_changing_frame_size() {
    ::test_changing_frame_size(&_pd, &_pt, &_synth, &_sf);
}

// ---------------------------------------------------------------------------
//	TestLorisSynthesis
// ---------------------------------------------------------------------------
void TestLorisSynthesis::setUp() {
    _sf = SndfileHandle(TEST_AUDIO_FILE);

    if(_sf.error() > 0) {
        throw Exception(std::string("Could not open audio file: ") +
                        std::string(TEST_AUDIO_FILE));
    }
}

void TestLorisSynthesis::test_basic() {
    ::test_basic(&_pd, &_pt, &_synth, &_sf);
}
